### PR TITLE
nodejs build style

### PR DESCRIPTION
**Pull request by selfisekai**
intended mostly for (bleh) electron packages

### CODE REVIEW

**Comment by Chocimier** [comment]
I do not want it as long as it install precompiled binaries, i. e. until #21758 is done and build style uses that.

**Comment by fosslinux** [comment]
Can we revisit this?

**Comment by Anachron** [comment]
We should export NODE_ENV=production, no?

**Comment by fosslinux** [comment]
Ping

**Comment by the-maldridge** [comment]
Looks good to me.  Needs docs to explain the new template variables that are supported by this build style and then it can be merged.  Ideally templates that can benefit from this style would be moved over to it as well.

**Comment by selfisekai** [comment]
@the-maldridge I did some simple docs and merged all the changes from master (force-pushed, because I did `git pull --rebase` and the PR included like 10k commits from master), you can check it now.

**Comment by fosslinux** [comment]
You need to rebase, not merge.

File: Manual.md [hunk 1]
```diff
@@ -915,6 +915,15 @@ depend on additional packages. This build style does not install
 dependencies to the root directory, and only checks if a binary package is
 available in repositories.
 
+- `nodejs` For apps made in Node.js that get built with npm or yarn
```
Comment (the-maldridge) [inline Manual.md:918]: I think the explanation of variables could be better formatted as a list.  Look at the python section starting on line 1504.

**Review by the-maldridge — CHANGES_REQUESTED** [review]
And you should address the rebase issue.

File: Manual.md [hunk 2]
```diff
@@ -1616,6 +1619,23 @@ The following variables influence how Haskell packages are built:
 - `make_build_args`: This is passed as-is to `stack build ...`, so
   you can add your `--flag ...` parameters there.
 
+<a id="pkgs_nodejs"></a>
```
Comment (ericonr) [inline Manual.md:1622]: This needs to go on the table of contents at the start of the document.

File: Manual.md [hunk 3]
```diff
@@ -1616,6 +1619,23 @@ The following variables influence how Haskell packages are built:
...
+Node.js packages should be built with the `nodejs` build style, if possible.
+The `nodejs` build style takes care of downloading yarn/npm dependencies
+and running builds via package.json scripts.
+
+The following template variables influence how Node.js packages are built:
+- `nodejs_build_script`: the package.json script that builds the app files
+	(default: `build`).
+- `nodejs_install_script`: app installing script, which most probably 
+	doesn't even exist and you should redefine `do_install`
+	in your build script.
```
Comment (ericonr) [inline Manual.md:1634]:
```suggestion
- `nodejs_install_script`: app installing script, which is unlikely to exist.
Most of the time, you will have to define your own `do_install`.
```

File: Manual.md [hunk 4]
```diff
@@ -1616,6 +1619,23 @@ The following variables influence how Haskell packages are built:
...
+
+The following template variables influence how Node.js packages are built:
+- `nodejs_build_script`: the package.json script that builds the app files
+	(default: `build`).
+- `nodejs_install_script`: app installing script, which most probably 
+	doesn't even exist and you should redefine `do_install`
+	in your build script. 
+- `nodejs_packager`: whether the build script should use yarn or npm.
+	Set to `yarn` if the `yarn.lock` file exists, else it's set to `npm`
+- `NODE_ENV`: set to `production` by default
```
Comment (ericonr) [inline Manual.md:1637]: it should be exported in the build style, otherwise just setting it won't affect the build.

**Comment by ericonr** [comment]
To be honest, though, I'm not sure I entirely see the value in this... #23948 seems to have successfully completed CI even without having the build style there.

**Comment by fosslinux** [comment]
@ericonr uh, there is a nodejs build style commit in that pr...

**Comment by ericonr** [comment]
well, oops. Still, the template defines its own `do_build` and `do_install`.

**Pull request closed without merging**